graylog2-web-interface/src/components/event-definitions/event-definition-form/field-value-providers/LookupTableFieldValueProviderForm.jsx

```diff
@@ -0,0 +1,130 @@
+import naturalSort from 'javascript-natural-sort';
+
+import { Select } from 'components/common';
+import FormsUtils from 'util/FormsUtils';
+
+class LookupTableFieldValueProviderForm extends React.Component {
+  static propTypes = {
+    allFieldTypes: PropTypes.array.isRequired,
+    config: PropTypes.object.isRequired,
+    eventFields: PropTypes.object.isRequired,
+    lookupTables: PropTypes.array.isRequired,
+    onChange: PropTypes.func.isRequired,
+  };
+
+  static type = 'lookup-v1';
+
+  formatMessageFields = lodash.memoize(
+    (fieldTypes) => {
+      return fieldTypes
+        .sort((ftA, ftB) => naturalSort(ftA.name, ftB.name))
+        .map((fieldType) => {
+          return {
+            label: `${fieldType.name} – ${fieldType.value.type.type}`,
+            value: fieldType.name,
+          };
+        });
+    },
+    fieldTypes => fieldTypes.map(ft => ft.name).join('-'),
+  );
+
+  propagateChanges = (key, value) => {
+    const { config, onChange } = this.props;
+    const nextProviders = lodash.cloneDeep(config.providers);
+    const templateProvider = nextProviders.find(provider => provider.type === LookupTableFieldValueProviderForm.type);
+    templateProvider[key] = value;
+    onChange(Object.assign({}, config, { providers: nextProviders }));
+  };
+
+  handleChange = (event) => {
+    const { name } = event.target;
+    const value = FormsUtils.getValueFromInput(event.target);
+    this.propagateChanges(name, value);
+  };
+
+  handleSelectChange = (key) => {
+    return (nextLookupTable) => {
+      this.propagateChanges(key, nextLookupTable);
+    };
+  };
+
+  formatFields = (context) => {
+    const { allFieldTypes, eventFields } = this.props;
+
+    if (context === 'source') {
+      return this.formatMessageFields(allFieldTypes);
+    }
+
+    return Object.keys(eventFields)
+      .sort(naturalSort)
+      .map(fieldName => ({ label: fieldName, value: fieldName }));
+  };
+
+  formatLookupTables = (lookupTables) => {
+    return lookupTables.map(table => ({ label: table.title, value: table.name }));
+  };
+
+  render() {
+    const { config, lookupTables } = this.props;
+    const provider = config.providers.find(p => p.type === LookupTableFieldValueProviderForm.type);
+
+    return (
+      <Row className="row-sm">
+        <Col md={12}>
+          <FormGroup>
+            <ControlLabel>Take Lookup Table key from...</ControlLabel>
+            <Radio id="lookup-message-key-context"
+                   name="key_context"
+                   value="source"
+                   checked={provider.key_context === 'source'}
+                   onChange={this.handleChange}>
+              Source log message
+            </Radio>
+            <Radio id="lookup-event-key-context"
+                   name="key_context"
+                   value="event"
+                   checked={provider.key_context === 'event'}
+                   onChange={this.handleChange}>
+              Generated Event
```

Comment:
I think we cannot use fields from the generated event, unfortunately. Because the field spec is a map/object instead of a list of objects in the backend, there is no defined execution order of the field spec. That means if we take a field for the lookup table provider from the event, it might not be defined yet.

So I guess we can only use fields from the message at the moment. Otherwise we would have to change the backend structure so we have a defined field provider execution order.

Reply:
All I did was reading the documentation we have for the provider and also the fields the API supports. I can remove that from the frontend, but we should do the same in the API if that is not going to work.

Reply:
Yes, the idea was to support this but then I realized that it doesn't work because we use a map instead of a list of maps in the backends.

Yeah, I will prepare a PR to remove it from the API as well. It's a backwards incompatible change, but since we didn't have any UI for that yet to create a lookup-table field provider, no one should run into any issues. :smile:

Reply:
The alternative is to redo the backend structure to make it a list of field providers. The we can define a processing order for field providers. But that would require lots of changes in the UI, I guess. The user then also needs to be able to reorder the field spec in the UI and has to actually think about the order to make sure referencing previously defined fields works. And then the backends needs to validate that as well...

So I guess it's easier to not allow using fields from the event. Easier for the user and for the implementation. :smile_cat: